crates/oxc_linter/src/rules/eslint/no_useless_backreference.rs

```diff
@@ -0,0 +1,554 @@
+}
+
+impl<'ast> Visit<'ast> for RegexCollector<'ast> {
+    fn enter_node(&mut self, kind: RegExpAstKind<'ast>) {
+        let id = self.nodes.push(kind);
+        self.stack.push(id);
+
+        match kind {
+            RegExpAstKind::CapturingGroup(group) => {
+                self.groups.push(GroupInfo {
+                    name: group.name.map(|n| n.as_str()),
+                    span: group.span,
+                    path: self.current_path(),
+                });
+            }
+            RegExpAstKind::IndexedReference(idx_ref) => {
+                self.backrefs.push(BackRefInfo {
+                    target: BackRefInfoTarget::Index(idx_ref.index),
+                    span: idx_ref.span,
+                    path: self.current_path(),
+                });
+            }
+            RegExpAstKind::NamedReference(named) => {
+                self.backrefs.push(BackRefInfo {
+                    target: BackRefInfoTarget::Name(named.name.as_str()),
+                    span: named.span,
+                    path: self.current_path(),
+                });
+            }
+            _ => {}
+        }
+    }
+
+    fn leave_node(&mut self, _kind: RegExpAstKind<'ast>) {
+        self.pop();
+    }
+}
+
+/// Determines if a backreference is useless — meaning it references a capturing
+/// group that hasn't (and cannot) participate at the time it's evaluated.
+//////
+/// Returns the reason as a `Problem` if one is found; otherwise returns `None`.
+///
+/// # Problem types and examples:
+///
+/// - `Problem::Nested`:
+///   `/ (a\1) /`
+///   ↳ Backref appears *inside* the group it references — it can't match because the group is still open.
+///
+/// - `Problem::Disjunctive`:
+///   `/ ((a)|\1b) /`
+///   ↳ Backref and group are in different alternatives — only one branch runs, so the group can't be matched.
+///
+/// - `Problem::Backward`:
+///   `/ \1(a) /`
+///   ↳ Backref appears before the group. Left-to-right matching means the group hasn’t matched yet.
+///
+/// - `Problem::Forward`:
+///   `/ (?<=\1(a)) /`
+///   ↳ In a lookbehind (right-to-left match), the group appears *after* the reference — invalid.
+///
+/// - `Problem::IntoNegativeLookaround`:
+///   `/ \1(?!(a)) /`
+///   ↳ Group is inside a negative lookaround — by definition, it can never successfully match.
+fn problems_for_backref<'a>(
+    bref: &'a BackRefInfo<'a>,
+    nodes: &IndexVec<RegexNodeId, RegExpAstKind<'a>>,
+    groups: &'a [GroupInfo<'a>],
+) -> Option<(
+    Problem,
+    /* span of the backreference */ Span,
+    /* span of the capture group */ Span,
+)> {
+    let cap_group = match bref.target {
+        BackRefInfoTarget::Index(i) => groups.get(i as usize - 1),
+        BackRefInfoTarget::Name(cap_group_name) => {
+            groups.iter().find(|g| g.name.is_some_and(|n| n == cap_group_name))
+        }
+    };
+    let Some(cap_group) = cap_group else {
+        // TODO: is this right? what happens when we can't map back ref to a capture group
```

Comment:
This can be noted as `unreachable!`. 🙆🏻 

In case of this, regexp parser does not mark `\1` as a reference, this is just an octal escape.

Reply:
awesome, thanks!